File: Sources/HedosKernel/Artifacts/Provenance.swift
import Foundation

public enum Provenance {
    public static func line(for artifact: Artifact, schema: [ParamSpec] = []) -> String {
        var parts = [artifact.model]
        parts.append(
            contentsOf: paramPairs(artifact.params, schema: schema).map { "\($0.key) \($0.value)" })
        parts.append(duration(ms: artifact.durationMs))
        return parts.joined(separator: " · ")
    }

    public static func duration(ms: Int) -> String {
        if ms < 1000 {
            return "\(ms) ms"
        }
        if ms < 60_000 {
            return String(format: "%.1fs", Double(ms) / 1000)
        }
        let minutes = ms / 60_000
        let seconds = (ms % 60_000) / 1000
        return seconds == 0 ? "\(minutes)m" : "\(minutes)m \(seconds)s"
    }

    public static func details(for artifact: Artifact, schema: [ParamSpec] = []) -> String {
        var lines = [
            "model: \(artifact.model)",
            "runtime: \(artifact.runtime)",
            "capability: \(artifact.capability.rawValue)",
        ]
        lines.append(contentsOf: promptAndParamLines(artifact.params, schema: schema))
        lines.append("duration: \(duration(ms: artifact.durationMs))")
        lines.append("job: \(artifact.jobID)")
        return lines.joined(separator: "\n")
    }

    public static func failureDetails(
        model: String, error: String, jobID: String?, params: JSONValue, schema: [ParamSpec] = []
    ) -> String {
        var lines = ["model: \(model)", "error: \(error)"]
        if let jobID {
            lines.append("job: \(jobID)")
        }
        lines.append(contentsOf: promptAndParamLines(params, schema: schema))
        return lines.joined(separator: "\n")
    }

    public static func prompt(of params: JSONValue) -> String? {
        guard case .object(let fields) = params, case .string(let prompt) = fields["prompt"]
        else { return nil }
        return prompt
    }

    private static func promptAndParamLines(_ params: JSONValue, schema: [ParamSpec]) -> [String] {
        var lines: [String] = []
        if let prompt = prompt(of: params) {
            lines.append("prompt: \(prompt)")
        }
        lines.append(
            contentsOf: paramPairs(params, schema: schema).map { "\($0.key): \($0.value)" })
        return lines
    }

    static func paramPairs(
        _ params: JSONValue, schema: [ParamSpec]
    ) -> [(key: String, value: String)] {
        guard case .object(let fields) = params else { return [] }
        var keys = schema.map(\.key).filter { fields[$0] != nil }
        let extras = fields.keys
            .filter { $0 != "prompt" && !keys.contains($0) }
            .sorted()
        keys.append(contentsOf: extras)
        return keys.compactMap { key in
            guard key != "prompt", let value = fields[key], let rendered = scalar(value) else {
                return nil
            }
            return (key, rendered)
        }
    }

    static func scalar(_ value: JSONValue) -> String? {
        switch value {
        case .int(let raw): String(raw)
        case .double(let raw): String(format: "%g", raw)
        case .string(let raw): raw
        case .bool(let raw): String(raw)
        default: nil
        }
    }
}

File: kernel/src/artifacts/provenance.rs
//! Rendering an artifact's provenance into a one-line summary and a multi-line
//! detail form. Known schema params come first (in schema order), then every
//! remaining scalar param alphabetically; `prompt` is pulled out separately.

use crate::records::{JsonValue, ParamSpec};

use super::artifact::Artifact;

/// Provenance rendering helpers.
pub struct Provenance;

impl Provenance {
    /// A one-line summary: model, the scalar params, and the duration.
    pub fn line(artifact: &Artifact, schema: &[ParamSpec]) -> String {
        let mut parts = vec![artifact.model.clone()];
        for (key, value) in param_pairs(&artifact.params, schema) {
            parts.push(format!("{key} {value}"));
        }
        parts.push(Self::duration(artifact.duration_ms));
        parts.join(" · ")
    }

    /// A human duration: `"850 ms"`, `"1.2s"`, `"3m"`, or `"3m 5s"`.
    pub fn duration(ms: i64) -> String {
        if ms < 1000 {
            return format!("{ms} ms");
        }
        if ms < 60_000 {
            return format!("{:.1}s", ms as f64 / 1000.0);
        }
        let minutes = ms / 60_000;
        let seconds = (ms % 60_000) / 1000;
        if seconds == 0 {
            format!("{minutes}m")
        } else {
            format!("{minutes}m {seconds}s")
        }
    }

    /// A multi-line detail form: model, runtime, capability, prompt, params, then
    /// duration and job.
    pub fn details(artifact: &Artifact, schema: &[ParamSpec]) -> String {
        let mut lines = vec![
            format!("model: {}", artifact.model),
            format!("runtime: {}", artifact.runtime),
            format!("capability: {}", artifact.capability.as_ref()),
        ];
        lines.extend(prompt_and_param_lines(&artifact.params, schema));
        lines.push(format!(
            "duration: {}",
            Self::duration(artifact.duration_ms)
        ));
        lines.push(format!("job: {}", artifact.job_id));
        lines.join("\n")
    }

    /// A detail form for a failed generation.
    pub fn failure_details(
        model: &str,
        error: &str,
        job_id: Option<&str>,
        params: &JsonValue,
        schema: &[ParamSpec],
    ) -> String {
        let mut lines = vec![format!("model: {model}"), format!("error: {error}")];
        if let Some(job_id) = job_id {
            lines.push(format!("job: {job_id}"));
        }
        lines.extend(prompt_and_param_lines(params, schema));
        lines.join("\n")
    }

    /// The `prompt` string of a params object, if present.
    pub fn prompt(params: &JsonValue) -> Option<String> {
        params
            .as_object()
            .and_then(|fields| fields.get("prompt"))
            .and_then(JsonValue::as_str)
            .map(str::to_owned)
    }
}

fn prompt_and_param_lines(params: &JsonValue, schema: &[ParamSpec]) -> Vec<String> {
    let mut lines = Vec::new();
    if let Some(prompt) = Provenance::prompt(params) {
        lines.push(format!("prompt: {prompt}"));
    }
    for (key, value) in param_pairs(params, schema) {
        lines.push(format!("{key}: {value}"));
    }
    lines
}

fn param_pairs(params: &JsonValue, schema: &[ParamSpec]) -> Vec<(String, String)> {
    let Some(fields) = params.as_object() else {
        return Vec::new();
    };
    let mut keys: Vec<String> = schema
        .iter()
        .map(|spec| spec.key.clone())
        .filter(|key| fields.contains_key(key))
        .collect();
    let mut extras: Vec<String> = fields
        .keys()
        .filter(|key| *key != "prompt" && !keys.contains(key))
        .cloned()
        .collect();
    extras.sort();
    keys.append(&mut extras);
    keys.into_iter()
        // The schema may itself list "prompt"; it is rendered on its own line.
        .filter(|key| key != "prompt")
        .filter_map(|key| {
            let value = fields.get(&key)?;
            let rendered = scalar(value)?;
            Some((key, rendered))
        })
        .collect()
}

fn scalar(value: &JsonValue) -> Option<String> {
    match value {
        JsonValue::Int(raw) => Some(raw.to_string()),
        JsonValue::Double(raw) => Some(raw.to_string()),
        JsonValue::String(raw) => Some(raw.clone()),
        JsonValue::Bool(raw) => Some(raw.to_string()),
        _ => None,
    }
}
